index.js

```diff
@@ -339,3 +339,71 @@ module.exports.random = {
     return s * Math.pow(2, (-64 - e))
   }
 }
+
+/**
+ * A dictionary of header values, i.e '{ 'header-name': 'header-value' }'
+ * @typedef {Object.<string, (string | string[] | undefined)>} HeaderLike
+ */
+
+/**
+ * Uses Ed25519: a public-key signature system {@link https://ed25519.cr.yp.to/} {@link https://datatracker.ietf.org/doc/html/draft-cavage-http-signatures-12}
+ * Signs the message using the secret key and returns a signature.
+ * @see {nacl.sign.detached}
+ *
+ * @method
+ * @param {string} keyId - an opaque string that the server can use to look up the component they need to validate the signature
+ * @param {string} secretKey - hex encoded secret key to sign the message
+ * @param {HeaderLike} headers - headers containing the properties to sign
+ * @returns {string}
+ */
+module.exports.ed25519HttpSign = function (keyId /* string */, secretKey /* string */, headers = {} /* dictionary */) {
+  if (!secretKey) throw new Error('secret key is required')
+  if (!keyId) throw new Error('key id is required')
+  if (Object.keys(headers).length === 0) throw new Error('headers are required')
+
+  const headerKeys = []
+  const message = Object.entries(headers)
+    .map(([key, value]) => {
+      headerKeys.push(key)
+      return `${key}: ${value}`
+    }).join('\n')
+
+  // Generate signature using the ed25519 algorithm.
+  const sign = nacl.sign.detached(
+    Uint8Array.from(Buffer.from(message)),
+    Uint8Array.from(Buffer.from(secretKey, 'hex'))
+  )
+
+  const signature = Buffer.from(sign).toString('base64')
+  return `keyId="${keyId}",algorithm="ed25519",headers="${headerKeys.join(' ')}",signature="${signature}"`
+}
+
+/**
+ * Uses Ed25519: a public-key signature system {@link https://ed25519.cr.yp.to/} {@link https://datatracker.ietf.org/doc/html/draft-cavage-http-signatures-12}
+ * Verifies the signature for the message and returns true if verification succeeded or false if it failed.
+ * @see {nacl.sign.detached.verify}
+ *
+ * @method
+ * @param {string} publicKey - hex encoded public key to verify the signature
+ * @param {HeaderLike} headers - headers containing the signature for verification
+ * @returns {boolean}
+ */
+module.exports.ed25519HttpVerify = function (publicKey /* string */, headers = {} /* dictionary */) {
+  if (!publicKey) throw new Error('public key is required')
+  if (!headers.signature) throw new Error('header signature is required')
+
+  const signedRequest = headers.signature.split(',').reduce((result, part) => {
+    const [key, value] = part.split('=')
+    if (value) result[key] = value.replace(/"/g, '') // remove quotes
+    return result
+  }, {})
+
+  if (!headers.signature) throw new Error('no signature was parsed')
```

Comment:
headers.signature contains in theory any arbitrary vale, this checks the full signature string version

Reply:
i'm a bit confused since it seems headers.signature has not been modified since https://github.com/brave/crypto/pull/72/files#diff-e727e4bdf3657fd1d798edcd6b099d6e092f8573cba266154583a746bba0f346R393. is this supposed to check signedRequest.signature instead?

Reply:
it seems to me that here it should check that signedRequest contains signature, headers, keyId, and `algorithm="ed25519"`, otherwise a required field is missing

Reply:
although this is the wrong variable 🤦

Reply:
My comments did not refresh, so apologies as you are right. I am fixing this and will check against the correct value